crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+        generator: &mut schemars::gen::SchemaGenerator,
+    ) -> openapiv3::v3_1::Operation {
+        OperationBuilder::new()
+            .tag("Transactions")
+            .operation_id("ResolveTransaction")
+            .request_body(
+                RequestBodyBuilder::new()
+                    // .json_content::<UnresolvedTransaction>(generator)
+                    .build(),
+            )
+            .response(
+                200,
+                ResponseBuilder::new()
+                    .json_content::<ResolveTransactionResponse>(generator)
+                    .bcs_content()
+                    .build(),
+            )
+            .build()
+    }
+
+    fn handler(&self) -> RouteHandler<RestService> {
+        RouteHandler::new(self.method(), resolve_transaction)
+    }
+}
+
+async fn resolve_transaction(
+    State(state): State<RestService>,
+    accept: AcceptFormat,
+    Json(unresolved_transaction): Json<UnresolvedTransaction>,
+) -> Result<ResponseContent<ResolveTransactionResponse>> {
+    let executor = state
+        .executor
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("No Transaction Executor"))?;
+    let (reference_gas_price, protocol_config) = {
+        let system_state = state.reader.get_system_state_summary()?;
+
+        let current_protocol_version = state.reader.get_system_state_summary()?.protocol_version;
+
+        let protocol_config = ProtocolConfig::get_for_version_if_supported(
+            current_protocol_version.into(),
+            state.reader.inner().get_chain_identifier()?.chain(),
+        )
+        .ok_or_else(|| {
+            RestError::new(
+                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                "unable to get current protocol config",
+            )
+        })?;
+
+        (system_state.reference_gas_price, protocol_config)
+    };
+    let called_packages =
+        called_packages(&state.reader, &protocol_config, &unresolved_transaction)?;
+    let user_provided_budget = unresolved_transaction
+        .gas_payment
+        .as_ref()
+        .and_then(|payment| payment.budget);
+    let mut resolved_transaction = resolve_object_references(
+        &state.reader,
+        &called_packages,
+        reference_gas_price,
+        protocol_config.max_tx_gas(),
+        unresolved_transaction,
+    )?;
+
+    let simulation_result = executor
+        .simulate_transaction(resolved_transaction.clone())
+        .map_err(anyhow::Error::from)?;
+
+    let budget = if let Some(user_provided_budget) = user_provided_budget {
+        user_provided_budget
+    } else {
+        let estimate = estimate_gas_budget_from_gas_cost(
+            simulation_result.effects.gas_cost_summary(),
+            reference_gas_price,
+        );
+        resolved_transaction.gas_data_mut().budget = estimate;
+        estimate
+    };
+
+    // If we ended up using a mock gas coin we need to do gas selection now
+    if simulation_result.mock_gas_id.is_some() {
```

Comment:
this can be determined without simulation (is payment unset)

Reply:
This is a place where the SDKs and dryRun diverged a bit.  Setting an empty payment list for dryRun was the only way to not specify gas for dryRun because we needed to resolve to bcs.

In the SDK, we have generally only done gas selection when the payment was unset rather just empty.  I am not sure how crucial this distinction is, but it would technically be a breaking change.

Reply:
Would it be okay to do the permissive thing on both sides? I.e. support both the case where the gas payment is not present at all, and the case where it is present but empty? I can't imagine someone setting the payment to an empty array and hoping that the dry-run would fail...

Reply:
yep will do!